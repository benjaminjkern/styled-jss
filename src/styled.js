import {Component, createElement} from 'react'
import PropTypes from 'prop-types'
import {themeListener, channel} from 'theming'

import filterProps from './utils/filterProps'
import composeClasses from './utils/composeClasses'
import generateTagName from './utils/generateTagName'
import getSeparatedStyles from './utils/getSeparatedStyles'

import type {
  JssSheet,
  ComponentStyleType,
  StyledElementPropsType
} from './types'

type StyledArgs = {
  tagName: string,
  elementStyle: ComponentStyleType,
  mountSheet: Function
}

type StateType = {
  theme?: Object
}

const styled = ({tagName, elementStyle, mountSheet}: StyledArgs) => {
  const {dynamicStyle, staticStyle} = getSeparatedStyles(elementStyle)
  const staticTagName = staticStyle && generateTagName(tagName)

  const availableDynamicTagNames = []
  const classMap = {}

  return class StyledElement extends Component {
    static tagName: string = tagName
    static style: ComponentStyleType = elementStyle
    static contextTypes = {
      [channel]: PropTypes.object
    }

    props: StyledElementPropsType
    sheet: JssSheet
    state: StateType
    unsubscribe: ?Function

    dynamicTagName = ''
    setTheme = (theme: Object) => this.setState({theme})

    constructor(props: StyledElementPropsType, context: any) {
      super(props, context)
      if (!this.dynamicTagName && dynamicStyle) {
        this.dynamicTagName = availableDynamicTagNames.pop() || generateTagName(tagName)
      }

      this.state = context[channel] ? {theme: themeListener.initial(context)} : {}
    }

    componentWillMount() {
      this.sheet = this.sheet || mountSheet()
      const rulesIndex = this.sheet.rules.index
      const rulesTotal = rulesIndex.length

      if (staticStyle && !this.sheet.getRule(staticTagName)) {
        this.sheet.addRule(staticTagName, staticStyle)
      }

      if (!dynamicStyle) return

      if (!this.sheet.getRule(this.dynamicTagName)) {
        this.sheet.addRule(this.dynamicTagName, dynamicStyle)
      }

      classMap[this.dynamicTagName] = classMap[this.dynamicTagName] || rulesIndex.slice(rulesTotal)
      this.updateSheet(this.props, this.state)
    }

    componentWillUpdate(nextProps: StyledElementPropsType, nextState: StateType) {
      if (dynamicStyle) this.updateSheet(nextProps, nextState)
    }

    componentDidMount() {
      if (this.state.theme) {
        this.unsubscribe = themeListener.subscribe(this.context, this.setTheme)
      }
    }

    componentWillUnmount() {
      availableDynamicTagNames.push(this.dynamicTagName)

      if (typeof this.unsubscribe === 'function') this.unsubscribe()
    }

    updateSheet(props: StyledElementPropsType, state: StateType) {
      let rule
      let ruleIndex = 0

      const styles = Object.assign({}, props, state)

      // nested styles become to flatten rules, so we need to update each nested rule
      for (ruleIndex; ruleIndex < classMap[this.dynamicTagName].length; ruleIndex++) {
        rule = classMap[this.dynamicTagName][ruleIndex]

        if (rule.name) this.sheet.update(rule.name, styles)
        if (rule.rules) rule.rules.update(styles)
      }
    }

    render() {
      const {children, className, ...attrs} = this.props

      const props = filterProps(attrs)
      const tagClass = composeClasses([
        staticTagName && this.sheet.classes[staticTagName],
        this.dynamicTagName && this.sheet.classes[this.dynamicTagName],
        className
      ])

      return createElement(tagName, {...props, className: tagClass}, children)
    }
  }
}

export default styled
