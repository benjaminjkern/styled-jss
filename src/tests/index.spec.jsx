import 'react-dom'
import renderer from 'react-test-renderer'
import React from 'react'

import styled, {Styled, injectStyled} from '../'

import CreateApp from './App'

describe('base rendering tests', () => {
  it('renders correctly App with default styled', () => {
    const App = CreateApp(styled)
    const tree = renderer.create(<App />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly App with default Styled', () => {
    const customStyled = Styled({
      baseButton: {
        color: 'red',
      },
    })

    const App = CreateApp(customStyled)
    const tree = renderer.create(<App />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly App with injectStyled', () => {
    const customStyled = Styled({
      root: {
        fontSize: 16
      },
      baseButton: {
        color: 'red',
      },
    })

    const App = CreateApp(customStyled)
    const StyledApp = injectStyled(customStyled)(({classes}) => (
      <div className={classes.root}><App /></div>
    ))
    const tree = renderer.create(<StyledApp />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders nested compositions correctly', () => {
    const C1 = ({children, className}: { children: any, className?: string }) => (
      <div className={className}>{children}</div>
    )
    const C2 = styled(C1)({color: '#333'})
    const C3 = styled(C2)({padding: 3})
    const tree = renderer.create(<C3>Test</C3>).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
