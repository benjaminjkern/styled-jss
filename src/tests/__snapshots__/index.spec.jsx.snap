// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base rendering tests renders correctly App with default Styled 1`] = `
<div
  className="div-8-0-9"
>
  <header
    className="header-9-0-10"
  >
    <h1
      className="h1-12-0-11"
    >
      Title
    </h1>
  </header>
  <section
    className="section-10-0-12"
    data-name="content"
  >
    <button
      className="button-13-0-13"
    >
      primitive test
    </button>
    <button
      className="button-14-0-14"
    >
      dynamic primitive test
    </button>
  </section>
  <section
    className="section-11-0-15"
  >
    Another section
  </section>
</div>
`;

exports[`base rendering tests renders correctly App with default styled 1`] = `
<div
  className="div-1-0-1"
>
  <header
    className="header-2-0-2"
  >
    <h1
      className="h1-5-0-3"
    >
      Title
    </h1>
  </header>
  <section
    className="section-3-0-4"
    data-name="content"
  >
    <button
      className="button-6-0-5"
    >
      primitive test
    </button>
    <button
      className="button-7-0-6"
    >
      dynamic primitive test
    </button>
  </section>
  <section
    className="section-4-0-7"
  >
    Another section
  </section>
</div>
`;

exports[`base rendering tests renders correctly App with injectStyled 1`] = `
<div
  className="root-0-16"
>
  <div
    className="div-15-0-18"
  >
    <header
      className="header-16-0-19"
    >
      <h1
        className="h1-19-0-20"
      >
        Title
      </h1>
    </header>
    <section
      className="section-17-0-21"
      data-name="content"
    >
      <button
        className="button-20-0-22"
      >
        primitive test
      </button>
      <button
        className="button-21-0-23"
      >
        dynamic primitive test
      </button>
    </section>
    <section
      className="section-18-0-24"
    >
      Another section
    </section>
  </div>
</div>
`;
